---
title: "Getting started with VoxHunt"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Getting started with VoxHunt}
  %\VignetteEngine{knitr::rmarkdown}
  \usepackage[utf8]{inputenc}
---

<style type="text/css">
body{
  font-family: Helvetica;
  font-size: 12pt;
}
</style>

```{r, echo = FALSE, message=FALSE}
knitr::opts_chunk$set(message = FALSE, warning = FALSE)
```

This vignette gives you an introduction on how to use VoxHunt to swiftly explore the Developing Mouse Brain Atlas ISH data, find brain structure-specific markers and project organoid cells to spatial brain maps.

## Installation

```{r, results='hide', eval=F}
# install.packages('devtools')
devtools::install_github('immunogenomics/presto')
devtools::install_github('quadbiolab/voxhunt')
```
```{r, results='hide'}
library(voxhunt)
```


## Exploring the ABA

VoxHunt provides a number of convenient functions to explore the Allen Developing Mouse Brain Atlas data computationally using R. To do so, you first need to point VoxHunt to the ABA gene expression data, which you can download [here](https://data.mendeley.com/datasets/n6488nxzbh/1). Now, we'll point VoxHunt to the location of these files:  
```{r, eval=F, echo=T}
load_aba_data('~/path/to/data')
```
```{r, results='hide', echo=F}
load_aba_data('~/projects/VoxHunt/data/voxhunt_rds/')
```
Per default, VoxHunt will not load all of the data at once, but only when we require it. 

Now that we have the data, we can start exploring. For instance, you can plot the E13.5 and P14 mouse brain with annotated brain structures:
```{r, fig.height=2, fig.width=8}
p1 <- plot_annotation('E13')
p2 <- plot_annotation('P4', show_legend=T)
ggarrange(p1, p2, nrow=1)
```

We can also plot the expression of different genes over these voxel maps:

```{r, fig.height=4, fig.width=6}
genes <- c('NEUROD6', 'EOMES', 'DCN', 'DLX1', 'DLX2', 'GBX2', 'OTX2', 'GATA3', 'PAX8')
plot_expression('E15', genes = genes, newpage = F)
```

Here we plotted just the sagittal view with a maximum intensity projection for all voxels, sometimes this hides the expression of certain genes. VoxHunt also allows you to define the sections that will be used in the plot:

```{r, fig.height=1.2, fig.width=2}
plot_expression('E15', genes = 'DCN', slices = 5:10)
```

We can also plot the expression of multiple genes for many coronal sections at the same time:

```{r}
plot_expression(
    'E15', 
    genes = genes, 
    view = 'slice', 
    slices = c(4, 12, 24, 30, 35, 40), 
    newpage = FALSE
)
```
```{r, fig.height=10, fig.width=7}
print(p)
```


In addition to plotting well-known canonical markers, we can also perform DE to find specific markers for brain structures we are interested in:

```{r, fig.width=4, fig.height=1.2}
marker_df <- structure_markers('E15', annotation_level = 'custom_4')
hippo_markers <- marker_df %>% 
    filter(group=='hippocampus') %>% 
    top_n(2, auc)
p <- plot_expression('E15', genes=hippo_markers$gene, nrow=1)
print(p)
```


## Transcriptomic similarity maps

Ok, now to the really interesting part: mapping single cells transcriptomes. Projecting single cells to these spatial maps based on several hundred genes can be very informative about cell type composition of the organoid. First, we load an Seurat object. The example case we use here contains 2300 cells from cerebral organoids.

```{r}
data(example_seurat)
head(example_seurat@meta.data)
```

Before we start mapping the cells, we'll do some feature selection to get region specific genes from the ABA:

```{r}
regional_markers <- structure_markers('E13') %>%
    group_by(group) %>%
    top_n(10, auc) %>% 
    {unique(.$gene)}
head(regional_markers)
```

Now let's use these markers to map our single cells to the brain. The `group_name` parameter refers to a metadata column that groups the data into clusters or cell types.

```{r}
vox_map <- voxel_map(
    example_seurat, 
    stage = 'E13', 
    group_name = 'cluster', 
    genes_use = regional_markers
)
print(vox_map)
```


The `plot_map()` function lets us plot this voxel map in a number of different ways. Per default, it will show us a sagittal view of the projection for each of the groups:

```{r}
p <- plot_map(vox_map, nrow=1)
```
```{r, fig.height=2, fig.width=8}
print(p)
```


However, as we have seen with the expression plots, we can also slice this plot however we want to reveal structures that might be hidden in the sagittal projection. For instance, the highest correlation for the ventral neurons seem to be hidden somewhere inside the brain, so lets look at fewer sagittal sections: 


```{r}
p <- plot_map(vox_map, view='slice', slices=seq(1, 40, 4))
```
```{r, fig.height=12, fig.width=3}
print(p)
```

Finally, VoxHunt also allows you to explore similarity maps interactively in 3D with a little help of plotly:

```{r}
plot_map_3d(
    vox_map, 
    show_group = 'ctx_cerebral', 
    width = 800, 
    height = 600
)
```



